### PR TITLE
Spending Report: add save button

### PR DESCRIPTION
**Pull request by carkom**
Saved as a LocalPref. Let me know your thoughts!

### CODE REVIEW

**Comment by matt-fidd** [comment]
Works perfectly for me!

What do you reckon to an indicator to show when there are unsaved changes?

**Comment by Teprifer** [comment]
Those is very neat as is, cheers.

Could the time frame option be included? Not a huge deal if not since they're easy as to set vs filters but could enhance the dashboard aspect.

**Comment by carkom** [comment]
I've added a "saved" indicator and incorporated the time option to save as well.

**Comment by matt-fidd** [comment]
The behaviour of the indicator is inconsistent when there is no filter saved.

When no filter has ever been saved, the indicator shows "save"
When the filter has been saved, then cleared and saved, it shows "saved"

Could the button hide when there are no filters in use and show when there are some either saved, or available to save?

**Comment by carkom** [comment]
Yea, I see what you mean. Also need to consider if someone has saved the filter but now wants the graph to display with no filters. Another thing to consider is the the "Save" button is also saving the "time" selection. So if someone wants no filters but "average" time selection they should be able to save in this way. Should also be able to change time selection (with no filters applied) and again click save to keep the new time choice.

So while it seems inconsistent at first glance it's actually working as intended if you consider both the filter and the time selection as possible options to be saved.

**Comment by matt-fidd** [comment]
Got you! It's only a minor visual thing as once it's used once it'll never happen again; and people that don't use it will never know the difference.

What do you think to a title attribute on button to explain the purpose?

image

**Comment by matt-fidd** [comment]
I'm getting a fatal error when I load the reports page now.

image

File: packages/desktop-client/src/components/reports/reports/SpendingCard.tsx [hunk 1]
```diff
@@ -20,12 +21,17 @@ export function SpendingCard() {
   const categories = useCategories();
 
   const [isCardHovered, setIsCardHovered] = useState(false);
+  const [spendingReportFilter = ''] = useLocalPref('spendingReportFilter');
+  const [spendingReportTime = 'lastMonth'] = useLocalPref('spendingReportTime');
 
+  const parseFilter = spendingReportFilter && JSON.parse(spendingReportFilter);
```
Comment (matt-fidd) [inline packages/desktop-client/src/components/reports/reports/SpendingCard.tsx:27]: The error appears to be coming from here:

```js
SyntaxError: "[object Object]" is not valid JSON
    at JSON.parse (<anonymous>)
    at U6 (SpendingCard.tsx:27:52)
    at nw (react-dom.production.min.js:167:137)
    at hD (react-dom.production.min.js:290:337)
    at cD (react-dom.production.min.js:280:389)
    at kN (react-dom.production.min.js:280:320)
    at Gp (react-dom.production.min.js:280:180)
    at Cx (react-dom.production.min.js:271:88)
    at sD (react-dom.production.min.js:268:429)
    at I (scheduler.production.min.js:13:203)
```

**Comment by matt-fidd** [comment]
Never mind, I cleared all site data and it resolved itself. Must have been something left over from testing an earlier commit.

**Review by matt-fidd — APPROVED** [review]

File: packages/desktop-client/src/components/reports/reports/Spending.tsx [hunk 2]
```diff
@@ -37,8 +40,28 @@ export function Spending() {
...
+    'spendingReportFilter',
+  );
+  const [spendingReportTime = 'lastMonth', setSpendingReportTime] =
+    useLocalPref('spendingReportTime');
+
   const [dataCheck, setDataCheck] = useState(false);
-  const [mode, setMode] = useState('lastMonth');
+  const [mode, setMode] = useState(spendingReportTime);
+
+  const parseFilter = spendingReportFilter && JSON.parse(spendingReportFilter);
```
Comment (jfdoming) [inline packages/desktop-client/src/components/reports/reports/Spending.tsx:52]: Wondering if this is worth memoizing, seems like we probably don't need to compute it on every render

File: packages/desktop-client/src/components/reports/reports/Spending.tsx [hunk 3]
```diff
@@ -37,8 +40,28 @@ export function Spending() {
...
+  const filterSaved =
```
Comment (jfdoming) [inline packages/desktop-client/src/components/reports/reports/Spending.tsx:53]: Ditto, maybe we could only recompute this if the object ID changes. Or actually, maybe just have a separate `useState` for `isSaved`?

File: packages/desktop-client/src/components/reports/reports/SpendingCard.tsx (hunk 1 above)
Comment (jfdoming) [inline packages/desktop-client/src/components/reports/reports/SpendingCard.tsx:27]: Same question as above

**Review by jfdoming — APPROVED** [review]
Looks good! Left some non-blocking comments

**Comment by jfdoming** [comment]
Oh one other thought: in combination with Matiss' PR to allow adding multiple spending reports to the report page, maybe the filter settings should be report-tile-specific instead of local prefs that apply to all spending reports.

**Comment by matt-fidd** [comment]
@carkom for visibility, I've just spotted a bug with this in edge now that I've updated. #3134

**Pull request merged**